File: Data-structure/Trie/trie.py
from collections import defaultdict
class TrieNode:
    def __init__(self):
        self.word = False
        self.children = defaultdict(TrieNode)

class Trie:
    def __init__(self):
        self.root = TrieNode()

    # 단어 삽입
    def insert(self, word:str)->None:
        node = self.root
        for char in word:
            node = node.children[char]
        node.word = True

    # 단어가 존재하는지
    def search(self, word:str)->bool:
        node = self.root
        for char in word:
            if char not in node.children:
                return False
            node = node.children[char]
        return node.word

    # 해당 문자열로 시작하는 단어가 존재하는지
    def startsWith(self, prefix:str)->bool:
        node = self.root
        for char in prefix:
            if char not in node.children:
                return False
            node = node.children[char]
        return True

trie = Trie()
trie.insert("apple")
print(trie.search("apple"))   # True
print(trie.search("app"))     # False
print(trie.startsWith("app")) # True
trie.insert("app")
print(trie.search("app"))     # True

File: Data-structure/Trie/Trie.js
class Node {
  constructor(value = "", stored = false) {
    this.value = value;
    this.children = new Map();
    this.stored = stored;
    this.searchedCount = 0;
  }
}

class Trie {
  constructor() {
    this.root = new Node();
  }

  insert(string) {
    let currentNode = this.root;

    for (const char of string) {
      if (!currentNode.children.has(char)) {
        currentNode.children.set(char, new Node(currentNode.value + char));
      }
      currentNode = currentNode.children.get(char);
    }
    // 검색된 단어 데이터 업데이트
    currentNode.stored = true;
    currentNode.searchedCount += 1;
  }

  has(string) {
    let currentNode = this.root;

    for (const char of string) {
      if (!currentNode.children.has(char)) {
        return false;
      }

      currentNode = currentNode.children.get(char);
    }
    return true;
  }

  autoComplete(string) {
    if (typeof string !== "string") throw new Error("인자가 문자열이 아니에요");
    if (!string.trim()) return null;

    const result = new Map();

    function dfs(node) {
      if (!node) return;
      const { value, children, stored, searchedCount } = node;
      if (stored) result.set(value, searchedCount);
      [...children.keys()].forEach((char) => dfs(children.get(char)));
      return;
    }

    let node = this.root;
    for (const char of string) {
      const { children } = node;
      if (!children.get(char)) return null;
      node = children.get(char);
    }

    dfs(node);

    // 검색이 많이된 순서대로 반환
    return [...result].sort((a, b) => b[1] - a[1]);
  }
}
